### ISSUE TITLE
Questions During Test Update PR Generation

### ISSUE THREAD

**Issue by billdenney** [nlmixr2/rxode2#17]
I'm currently working on a test update to switch to testthat edition 3 and to try to make the tests silent when successful.

During that, I have a few questions:

1. Similar to the suggestion in nlmixr2, I think it would be helpful and simplifying to switch the testing to remove the rxode2Test function.  Do you agree?
2. There are a bunch of calls like `.rx <- loadNamespace("rxode2")`.  Would it be okay to switch all of those to be `rxode2:::` instead of `.rx$` to make it a bit clearer what is happening in the tests?  But, I'm not certain that all the `.rx$` calls are all just doing the same as `:::`.  I don't think it could be a search-and-replace simple switch.
3. Many of the tests reset the random number generator with set.seed().  I think that it would be better to use `withr::with_seed()`.  Do you agree?
4. Longer-term, I think it would be helpful to move the tests like `test-rxode-issue-050.R` to be in a test file that points to the function being tested.  Do you agree with that one?
5. Several tests suggest that they require the `units` library.  Do you know if the units library is really required?
6. When running tests in parallel, they tend to stop because of "rxode2 already building model, waiting for lock file removal".  Is there a way to make this not an error or less strict?  I can imagine using nlmixrtargets in a way where multiple models may be built simultaneously, too.  (Though, I've also had some places where I've made errors in updating the testing, so maybe this is a non-issue.)

**Comment by mattfidler** [comment]
> Similar to the suggestion in nlmixr2, I think it would be helpful and simplifying to switch the testing to remove the rxode2Test function. Do you agree?

Yes

> There are a bunch of calls like .rx <- loadNamespace("rxode2"). Would it be okay to switch all of those to be rxode2::: instead of .rx$ to make it a bit clearer what is happening in the tests? But, I'm not certain that all the .rx$ calls are all just doing the same as :::. I don't think it could be a search-and-replace simple switch.

This was done to reduce the instances of `:::` in the package.  It is an error in the `codefactor` static code analysis.  I would prefer it to be left as is because it allows detection of such issues in the main code base while keeping the `:::` errors to a minimum

> Many of the tests reset the random number generator with set.seed(). I think that it would be better to use withr::with_seed(). Do you agree?

I agree with the concept, but `withr::set_seed()` is not sufficient.  `rxode2` has its own internal seed generation mechanism that should be saved and restored, so this should set and restore the `R` and `rxode2` seed.

> Longer-term, I think it would be helpful to move the tests like test-rxode-issue-050.R to be in a test file that points to the function being tested. Do you agree with that one?

I think this is fine.  I would also keep the issue in the comments somewhere.

> Several tests suggest that they require the units library. Do you know if the units library is really required?

Yes.  It is used in the event table.  You can convert units.  In theory this could be extended somehow to create automatic scaling factors in nlmixr.

> When running tests in parallel, they tend to stop because of "rxode2 already building model, waiting for lock file removal". Is there a way to make this not an error or less strict? I can imagine using nlmixrtargets in a way where multiple models may be built simultaneously, too. (Though, I've also had some places where I've made errors in updating the testing, so maybe this is a non-issue.)

Maybe it can simply wait in steading erroring out.  Without it parallel `rxode2` has caused issues in parallel nlmixr fits.

**Comment by john-harrold** [comment]
> > When running tests in parallel, they tend to stop because of "rxode2 already building model, waiting for lock file removal". Is there a way to make this not an error or less strict? I can imagine using nlmixrtargets in a way where multiple models may be built simultaneously, too. (Though, I've also had some places where I've made errors in updating the testing, so maybe this is a non-issue.)
> 
> Maybe it can simply wait in steading erroring out. Without it parallel `rxode2` has caused issues in parallel nlmixr fits.

Is this because you create a lot of objects that the ode solver is using for the model and these have standard names? I ran into this issue in ubiquity. To avoid it I created a md5sum hash of the model file name (I store models in text files) and then appended that md5sum to each object related to that model. Then I just made sure to append that sum to the objects I was looking for.

**Comment by billdenney** [comment]
@mattfidler , The PR #16 won't address the items listed here.  That PR covers items that aren't here.

**Comment by mattfidler** [comment]
> Is this because you create a lot of objects that the ode solver is using for the model and these have standard names? I ran into this issue in ubiquity. To avoid it I created a md5sum hash of the model file name (I store models in text files) and then appended that md5sum to each object related to that model. Then I just made sure to append that sum to the objects I was looking for.

This is a good point.  Last year there was a similar issue with RxODE and C-based symbols.  All items are now unique.  This was strictly enforced by CRAN.  Currently the "common" model components are translated to rx md5 and item number.

The problem here is multiple processes are trying to compile the same model (I think).  Maybe Bill can give more information.

**Comment by john-harrold** [comment]
You could add some salt to the models in each test to make them generate a unique checksum so you dont run into this issue.

**Comment by mattfidler** [comment]
> You could add some salt to the models in each test to make them generate a unique checksum so you dont run into this issue.

Perhaps, but then you couldn't load a pre-compiled model either...

**Comment by billdenney** [comment]
I haven't been able to replicate the issue as it occurred while I was reworking the testing, and I had some errors that were stopping the testing in the middle.

It seemed like it was trying to recompile the same model multiple times, so I agree with @mattfidler 's assessment.  And, I think that the reason it was trying to do that was due to stopping the testing partway through.

I don't think that this is something that needs a fix (since I don't think we need to go down the path of extreme lock testing).

**Comment by billdenney** [comment]
The only item not covered in the list is now moving tests like `test-rxode-issue-050.R` to other places.  Since that is so low priority, I'm going to close this now.
